esp/scm/ws_config2.ecm

```diff
@@ -0,0 +1,368 @@
+};
+
+
+ESPresponse ListOpenSessionsResponse
+{
+    bool   error(false);
+    string msg("");
+    ESParray<ESPstruct openSessionInfo, sessionInfo> openSessions;
+};
+
+
+//
+// Node management
+
+ESPrequest NodeRequest
+{
+    string sessionId;
+    string nodeId;
+};
+
+
+ESPrequest RemoveNodeRequest
+{
+    string sessionId;
+    string sessionLockKey("");
+    string nodeId;
+};
+
+
+ESPstruct choiceType
+{
+    string displayName;
+    string value;
+    string desc("");
+};
+
+
+ESPstruct limitsType
+{
+    bool   minValid(false);
+    bool   maxValid(false);
+    int    min(0);
+    int    max(0);
+    ESParray<ESPstruct choiceType, choice> choiceList;
+    //ESParray<string, disallow> disallowList;
+    ESParray<string, expr> regex;
+};
+
+
+ESPstruct typeInfo
+{
+    string name("na");
+    ESPstruct limitsType limits;
+    ESParray<string, modifier> modifiers;
+};
+
+
+ESPstruct nodeInfoType
+{
+    string    name("");
+    string    nodeType("");
+    string    class("");
+    string    category("");
+    bool      required(false);   // only used when in the insertable list
+    bool      hidden(false);
+    string    tooltip("");
+};
+
+
+ESPstruct nodeType
+{
+    string    nodeId;
+    int       numChildren(0);
+    ESPstruct nodeInfoType nodeInfo;
+};
+
+
+ESPstruct attributeType
+{
+    string     displayName;
+    string     name;
+    string     tooltip("");
+    ESPstruct  typeInfo type;
+    string currentValue("");
+    string defaultValue("");
+    bool   required(true);
+    bool   readOnly(false);
+    bool   hidden(false);
+};
+
+
+ESPstruct attributeValueType
+{
+    string    name;
+    string    value;
+    bool      force(false);  // set to true to allow setting a non-legal value, or to add a new value that is not defined
+};
+
+
+ESPresponse [exceptions_inline, encode(0)] GetNodeResponse
+{
+    string nodeId;
+    bool   error(false);
+    string msg("");
+    ESPstruct nodeInfoType nodeInfo;
+    ESPstruct statusType status;
+    ESParray<ESPstruct attributeType, attribute> attributes;
+    ESParray<ESPstruct nodeType, child> children;
+    ESParray<ESPstruct nodeInfoType, item> insertable;           // list of insertable elements under this node
+    bool localValueDefined(false);                               // true if the node is configured to have a value (not usually true)
+    ESPstruct attributeType value;                               // this would be <element>value</element> not normal, see valueDefined
+};
+
+
+ESPrequest InsertNodeRequest  // response is same as GetPathRequest
+{
+    string sessionId("0");
+    string sessionLockKey("");
+    string parentNodeId;
+    string nodeType;  // name of new instance to create
+};
+
+
+ESPrequest SetValuesRequest
+{
+    string sessionId("0");
+    string sessionLockKey("");
+    string nodeId;
+    ESParray<ESPstruct attributeValueType, attribute> attributeValues;
+    bool allowInvalid(false);  // true to allow setting an invalid value
```

Comment:
Interesting that the GUI gets to decide that an invalid value is allowed at a given location, rather than the schema or component owner.  Are we sure we want to set this as a permanent precedent?  It's sort of like a "force value" if we do keep it, "forceInvalid" may be a better name since its coming from the front end.  Allow seems like a backend definition or configuration thing to me.

Reply:
I'd like to discuss as this was a requirement conveyed to me in early meetings that the config manager essentially needed to allow any value to be set. I settled on adding a flag that allowed the situation. Also consider the case where the config requires several components. You may setup one that needs to reference something that isn't setup yet. In this state the config is not valid. However, you next configure another component that satisfies the dependency and now the config is valid.